docs/agent-development.md

```diff
@@ -384,3 +387,55 @@ Here's a list of the config options across all agents, their types, default valu
+2. Using the `POST` method, encode the parameters as a JSON object in the body, e.g. `{"service": {"name": "opbeans", "environment": "production"}}`
+
+The server will respond with a JSON object, where each key maps a config attribute to a string value. The string value should be interpreted the same as if it were passed in via an environment variable. Upon receiving these config changes, the agent will update its configuration dynamically, overriding any config previously specified. That is, config via Kibana takes highest precedence.
+
+To minimise the amount of work required by users, agents should aim to enable this feature by default. This excludes RUM, where there is a performance penalty.
+
+#### Caching
+
+As mentioned above, the server will cache config for each unique `service.name`, `service.environment` pair. The server will respond to config requests with two related response headers: [Etag](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/ETag) and [Cache-Control](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/Cache-Control).
+
+Agents should keep a record of the `Etag` value returned by the most successful config request, and then present it to future requests via the [If-None-Match](https://developer.mozilla.org/en-US/docs/Web/HTTP/Headers/If-None-Match) header. If the config has not changed, the server will respond with 304 (Not Modified).
+
+The `Cache-Control` header should contain a `max-age` directive, specifying the amount of time (in seconds) the response should be considered "fresh". Agents should use this to decide how long to wait before requesting config again. The server will respond with a `Cache-Control` header even if the request fails.
+
+#### Dealing with errors
+
+Agents must deal with various error scenarios, including:
+
+ - 7.3 servers where the Kibana connection is not enabled (server responds with 403)
+ - 7.3 servers where the Kibana connection is enabled, but unavailable (server responds with 503)
+ - any other error (server responds with 5xx)
+ - pre-7.3 servers that don't support the config endpoint (server responds with 404)
+
+If the server responds with 403, agents should log the response at info level.
+
+If the server responds with any 5xx, agents should log at error level. In any case, the server _should_ respond with a Cache-Control header, as described in the section above, and agents should retry after the specified interval. If for whatever reason the server does not respond with that header, or it is invalid, agents should retry after 5 minutes.
+
+If the server responds with 404 (i.e. it's a pre-7.3 server that lacks the config endpoint), agents may choose to treat it the same as 403, or may choose to log only once and mark that server as not having support for config, and then never retry. Bear in mind that this would mean the application must be restarted if the server is upgraded to 7.3.
```

Comment:
This is complicated by the fact that the agent could get a 404 for two different reasons: because there's no config _endpoint_, or because there's no service/environment pair. The agent should definitely not treat the latter case as an error

Reply:
> (i.e. it's a pre-7.3 server that lacks the config endpoint)

As you also pointed out, this is wrong. This whole paragraph should probably just be updated to say that it can happen both in the case that this is pre-7.3, or in the case that there's simply no config available currently.

Reply:
Fair enough. I guess it doesn't really matter anyway, since if they're using pre-7.3, it shouldn't be very surprising that there's no central config :)